src/model/mod.rs

```diff
@@ -48,8 +50,145 @@ pub enum PlayingItem {
     Episode(show::FullEpisode),
 }
 
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Id<'id> {
+    _type: Type,
+    id: &'id str,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Display, Error)]
+pub enum IdError {
+    InvalidPrefix,
+    InvalidFormat,
+    InvalidType,
+    InvalidId,
+}
+
+impl std::fmt::Display for Id<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str("spotify:")?;
+        f.write_str(self._type.as_ref())?;
+        f.write_str(":")?;
+        f.write_str(self.id)?;
+        Ok(())
+    }
+}
+
+impl AsRef<str> for Id<'_> {
+    fn as_ref(&self) -> &str {
+        self.id
+    }
+}
+
+impl Id<'_> {
+    pub fn _type(&self) -> Type {
+        self._type
+    }
+
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+
+    pub fn uri(&self) -> String {
+        format!("spotify:{}:{}", self._type.as_ref(), self.id)
+    }
+
+    pub fn from_id_or_uri<'a, 'b: 'a>(_type: Type, id_or_uri: &'b str) -> Result<Id<'a>, IdError> {
```

Comment:
IMHO, I think `from_id_or_uri` should be more explicit, just use this function to `get_id` or `get_uri`, instead of trying them both, it's a bit implicit.

Reply:
One of the problems it solves, it unifies URI and id arguments in methods, so a user can use either Spotify id or URI, no matter what real Spotify API endpoint requires. So this method accepts either id or URI from one of the API endpoint methods, hence the name.

Reply:
I think it's fine to have `from_id_or_uri` as long as we also have `from_id` and `from_uri`, just so that the user can call it as a convenience method for whenever they don't know what you're parsing (say, user input).